lib/core/src/Cardano/Wallet.hs

```diff
@@ -692,34 +696,40 @@ newWalletLayer tracer bp db nw tl = do
                         then Ready
                         else Restoring progress'
 
-            let txs_cps = applyBlocks @s @t (NE.toList blocks) cp
-            let txs = fold $ fst <$> txs_cps
-
-            let cpLast = fromMaybe cp $ lastMay $ snd <$> txs_cps
-            let (Quantity bhLast) = blockHeight cpLast
+            let putCheckpoint cp txs = do
+                    let meta' = calculateMetadata (view #slotId $ currentTip cp)
+                    DB.putCheckpoint db (PrimaryKey wid) cp
+                    DB.putTxHistory db (PrimaryKey wid) txs
+                    DB.putWalletMeta db (PrimaryKey wid) meta'
```

Comment:
Is this really correct?

While it is the case that we only need to write checkpoints for **unstable** blocks, we still need to write the TX history and metadata for **all** blocks, right?

Reply:
Yes but the mistake is actually a bit below line 717:

```hs
             putCheckpoint (snd cpLast) (fst cpLast) 
```

This should be:

```hs
             putCheckpoint (snd cpLast) newTxs
```

where `newTxs` is defined earlier as:

```hs
             let newTxs = fold $ fst <$> cps 
```

The meta only updates the progress, so we're fine updating only the last one for the last block applied (as we did before). Anyway, good catch, but I am slightly worried that our unit tests didn't quite catch that :no_mouth: ...

Reply:
Fixed.